### DISCUSSION TITLE
Dynamic routing: router vs getStaticProps(params)

### DISCUSSION THREAD

**Discussion by JLHasson** [vercel/next.js#17051] [Help]
In the [intro to next.js](https://nextjs.org/learn/basics/dynamic-routes/implement-getstaticprops) they use 
```javascript
export async function getStaticProps({ params }) {
  const postData = getPostData(params.id)
  ...
}
```
to access the `id` parameter from the dynamic route `/posts/[id].js`. In the documentation for [dynamic routes](https://nextjs.org/docs/routing/dynamic-routes) there's no mention of this technique, and instead they're using
```javascript
const router = useRouter()
const { id } = router.query
```
for some route like `/posts/[id].js`. Which one of these methods is preferred?

**Comment by uinstinct** [comment]
You have to use **getStaticPaths** to first fetch all the ids that you want in the dynamic route
Then, you have to use:
```js
export async function getStaticPaths() {
  return {
    paths: [
      { params: { id : "your ids" } } // populate this array using an api call or so
    ],
    fallback: false
  };
}
```

In the **getStaticProps(ctx)**, you will now get id in the _ctx.params_ object

**Reply by JLHasson** [reply]
I understand that method of doing things, but it looks to me like I can completely ignore all of the `getStaticPaths` code and just use 
```javascript
const router = useRouter()
const { id } = router.query
```
to access the info I need. So if I navigate to `/posts/1` then `const { id } = router.query // id = 1`. Is that correct or am I misunderstanding?

Is the difference that in one case the page content is totally statically generated (when using `getStaticPaths`) and the other case the page content is dynamically loaded via js (using the value of `id` retrieved from `router`)?

**Reply by uinstinct** [reply]
If you are not using `getStaticProps`, you have to look that `router.query.id` is not undefined in the `export default function` of the page or in the components you are passing _it_ as a prop (because it takes a second to get that `id`). 

I think `next/router` is available only on a page component. So, it makes sense to use `getStaticProps`
